Feature/tulcdm 141a ensure assets handled similar to restricted access

skng5 opened the pull request:
Ready for review. Merge if it looks good.
- Visibility of private collection is defined with a [literal](https://github.com/tulibraries/tul_cdm/blob/feature/TULCDM-141a-Ensure_assets_handled_similar_to_restricted_access/app/controllers/application_controller.rb#L22). Should we add a list of groups allowed access to the private collection? If so, then should it be part of this issue?
- In RSpec tests, cannot pass authenticated session to application_controller methods. Seems to work in development. Considering moving these methods into an ActiveSupport::Concern module.

nomadicoder committed:
Validate IP address with ipaddr library

- Also fix simple coverage false positive problem -- misses checking
  coverage in app/helpers

nomadicoder committed:
Add masked ip addresses and tests

nomadicoder committed:
Use proxy to govern viewability

- Updated viewability logic
- Refactor tests

nomadicoder committed:
Pre-holiday break commit

- Filter only in ip-address
- Begin testing on viewable, unviewable, and ip restricted collections

nomadicoder committed:
Fix filtering logic

- Private collection is viewable if an archivist is logged in
- Comperhensive tests

nomadicoder committed:
Comment for future fix

nomadicoder committed:
Merge branch 'develop' into feature/TULCDM-141a-Ensure_assets_handled_similar_to_restricted_access

Conflicts:
	spec/rails_helper.rb

nomadicoder committed:
Explain exception handling in is_visible?

bibliotechy commented on app/controllers/application_controller.rb:
```diff
@@ -12,12 +14,46 @@ class ApplicationController < ActionController::Base
+    if (collection['allowed_ip_addresses'].empty?)
+      if (is_private?(collection))
+        # TODO: Brittle - Allow better control over who may view a private collecionts
+        RoleMapper.roles(current_user.email).include? "archivist"
```
@skng5 I need a little help understanding this line. Why are we checking if the email of the current user contains the string "archivist"? Is that a default email address of some kind for admin users?

skng5 replied:
Email is passed to RoleMapper.roles method to get the Role mapping in the config/role_map.yml. current_user is the user that's logged into the app by way of http://locahost:3000/users/sign_in.
- Note - I updated the PR description with an additional item to discuss.

bibliotechy commented:
@skng5 OK, that makes sense. Thanks. And you note above it highlights my actual concern, I think.

bibliotechy commented:
@skng5 Trying to suss out why functions like `is_viewable?` and `unallowed_by_ip_collections` are in both applications_controller and tul_cdm_helper. It looks like a direct copy paste.

skng5 commented:
@bibliotechy : That was operator error. I started moving the code to the helper with the intention of getting it out of the application controller, but decided to keep it in application controller before I issued the PR. I forgot to revert the helper. I'd like to spend more time with this issue to research moving it out of the application controller into either the helper or a concern.

bibliotechy commented:
@skng5 That sounds like a good idea. Let's talk tomorrow about the scope of that.

nomadicoder committed:
Move visibility into its own concern

skng5 commented:
In  lib/tul_cdm/models/base.rb I added Hydra::AccessControls which controls show and search results in catatalog_controller with:

``` ruby
  # These before_filters apply the hydra access controls
  before_filter :enforce_show_permissions, :only=>:show
  # This applies appropriate access controls to all solr queries
  CatalogController.solr_search_params_logic += [:add_access_controls_to_solr_params]
```

But this requires us to set permission on ingested objects. This adds complexity to the ingest process. I would like to remove this add-on and write our own custom enforce_show_permissions and a add_access_controls_to_solr that uses the access control we've already created to gate access to Digital Collections.

nomadicoder committed:
Use only Digital Collection restrictions

- Remove HydraAccess::Control
- Modify catalog controller to use Digital Collection attributes
  to gate access to content
- Clean up test descriptions

nomadicoder committed:
Change spec to use new multi-ip private collection

nomadicoder committed:
Remove HydraAccess::Control code from ingest

nomadicoder committed:
Remove redundant/dead code

nomadicoder committed:
Unless a private object is visible, login

skng5 commented:
I've made the changes above. I've removed Hydra::AccessControl and it depends solely on access controls in the Digital Collections model. Please review and merge if you think it's ready.

skng5 commented:
To try this feature, use the `seed.rb` file, and set the `is_private` field to `true` of one of the collections.

bibliotechy commented:
@skng5 Where are IP address ranges for each collection coming from? Are we going to have to define them in a config / seed file? So every time it is updated on ContentDM, it will also have to be updated here? Seems likely to result in config drift between systems.

skng5 commented:
@bibliotechy The IP address ranges will come from the stakeholders. They can be set in the seed.rb, but best practices dictates that the information be entered and modified using the digital collection admin page. The seed should only be run once, when the system is initially started and the database backed up on a regular basis.
